src/write.rs

```diff
@@ -476,7 +483,177 @@ where
                 Absorb::absorb_second(w_inner, op, &*r_handle);
             }
         } else {
-            self.oplog.extend(ops);
+            // Only try to compress if it is enabled, else use efficient fallback.
+            if T::MAX_COMPRESS_RANGE > 0 {
+                // Compress oplog by rev-iterating all ops appended since the last publish
+                // while attempting to combine them with the next op,
+                // cut short when an attempt fails due to encountering a dependency (e.g. clear then set).
+
+                // used to avoid walking more of the oplog than necessary.
+                let mut rev_dirty_range = 0usize..0;
+                for mut next in ops {
+                    // used to avoid linear insertion time in case of predominantly independent ops.
+                    let mut range_remaining = T::MAX_COMPRESS_RANGE;
+                    // used to more efficiently insert next if possible
+                    let mut none: Option<(usize, &mut Option<O>)> = None;
+                    // rev-iterate all unpublished and potentially non-none ops already in the oplog
+                    for (prev_rev_idx, prev_loc) in {
+                        #[cfg(test)]
+                        {
+                            // While testing, make very very sure we don't skip any Some.
+                            let none_rev_idx = rev_dirty_range.start.saturating_sub(1);
+                            self.oplog
+                                .iter_mut()
+                                .skip(self.swap_index) // only consider the fresh part of the oplog
+                                .rev() // We need to walk it in reverse
+                                .enumerate() // we need the reverse index for the rev_dirty_range optimization
+                                .skip_while(move |(rev_idx, loc)| {
+                                    if *rev_idx < none_rev_idx {
+                                        debug_assert!(loc.is_none(), "We never skip over Some.",);
+                                        true
+                                    } else {
+                                        debug_assert!(
+                                            *rev_idx == none_rev_idx || loc.is_some(),
```

Comment:
This means that we don't check whether `non_rev_idx` is a `None` or a `Some`. Is that intentional?

Reply:
Yes. Because `let none_rev_idx = rev_dirty_range.start.saturating_sub(1);`, the first index we don't skip is None if there are any. It is used to avoid just pushing the op if compression fails.